### PR TITLE
Rename LambdaType to ArrowType

### PR DESCRIPTION
**Pull request by peter-evans**
Closes https://github.com/hylo-lang/hylo/issues/385

I did my best to change all references to "lambda" that aren't referring to *actual* lambda expressions. It wasn't immediately clear in some cases, so I might have made a mistake or missed some references.

I've included grep output below for convenience. Let me know if I've missed references that should be updated.

<details>
<summary>rg Lambda</summary>
<br />

```
rg Lambda
Sources/Core/AST/Expr/LambdaExpr.swift
2:public struct LambdaExpr: Expr {

Sources/Core/AST/Decl/SynthesizedFunctionDecl.swift
24:    /// Lambda generated for an autoclosure argument.

Sources/Core/AST/NodeIDs/NodeKind.swift
115:    LambdaExpr.self,

Sources/Core/AST/AST+Walk.swift
124:    case LambdaExpr.self:
125:      traverse(self[n] as! LambdaExpr, notifying: &o)
503:    _ n: LambdaExpr, notifying o: inout O

Sources/FrontEnd/Parse/Parser.swift
1679:      // Lambda expression.
1680:      return try parseLambdaExpr(in: &state).map(AnyExprID.init)
1962:  private static func parseLambdaExpr(in state: inout ParserState) throws -> LambdaExpr.ID? {
1983:      LambdaExpr(decl: decl, site: state.ast[decl].site))

Sources/FrontEnd/TypeChecking/TypeChecker.swift
2555:    of e: LambdaExpr.ID, withHint hint: ArrowType?
2593:    of e: LambdaExpr.ID, withHint hint: [CallableTypeParameter]?
4700:    case LambdaExpr.self:
4701:      return _inferredType(of: LambdaExpr.ID(e)!, withHint: hint, updating: &obligations)
4947:    of e: LambdaExpr.ID, withHint hint: AnyType? = nil,
5364:    of e: LambdaExpr.ID, withHint hint: AnyType? = nil,

Sources/IR/Emitter.swift
1413:    case LambdaExpr.self:
1414:      emitStore(LambdaExpr.ID(e)!, to: storage)
1597:  private mutating func emitStore(_ e: LambdaExpr.ID, to storage: Operand) {

Docs/MetatypeLowering.md
86:**Lambda types:**

Tests/HyloTests/ParserTests.swift
1275:  func testLambdaExpr() throws {
1278:    let expr = try XCTUnwrap(ast[exprID] as? LambdaExpr)
```
</details>

<details>
<summary>rg lambda</summary>
<br />

```
rg lambda
Sources/Core/Program.swift
459:        return qualification + ".lambda"

Sources/Core/AST/Expr/ParameterTypeExpr.swift
12:  /// `true` if arguments to the parameter are automatically wrapped in lambdas.

Sources/Core/AST/Expr/LambdaExpr.swift
1:/// A lambda.

Sources/FrontEnd/Parse/Parser.swift
1969:    let body = try state.expect("function body", using: lambdaBody)
1986:  private static let lambdaBody = inContext(
2131:      body = try state.expect("function body", using: lambdaBody)

Sources/FrontEnd/TypeChecking/TypeChecker.swift
890:  /// - Requires: `d` is not the underlying declaration of a lambda.
1094:  /// Type checks `d`, which declares a lambda, after its type has been inferred.
1096:    assert(program[d].isInExprContext, "expected lambda")
2548:  /// `hint` is used as contextual information to refine guesses if it is a lambda type with the
2565:    // Slow path: infer elided conventions from the uses in the lambda's body.
2903:  /// underlying declaration of a lambda). Otherwise, `.void` is returned.

StandardLibrary/Sources/Array.hylo
47:    // TODO: Call `self.copy()` directly in the lambda.

Docs/Subscripts.md
70:The caller initiates a projection by calling the ramp function and passing it a lambda representing at least the code using the projection.
71:This continuation is called with another lambda denoting the tail corresponding to the `yield` instruction executed by the ramp.
72:It is expected to call that lambda when it is done with the projection.

Docs/CPlusPlus.org
123:For interop purposes, projections could be translated into functions that accept a C++ lambda

Docs/MetatypeLowering.md
22:- Kind (e.g., product type, lambda type, etc.)
89:struct hylo_lambda_type_mirror {

Tools/gyb.py
320:            in tokenize.generate_tokens(lambda i=iter(source_lines):
367:            in tokenize.generate_tokens(lambda i=iter(source_lines): next(i)):

Tests/HyloTests/TestCases/Lowering/MemberLambda.hylo
4:// call to a member property with a lambda type.

Tests/EndToEndTests/TestCases/Lambda.hylo
8:  // Applies a thin, immutable lambda.
12:  // Applies a thick, immutable lambda.
17:  // Applies a mutable lambda with owned captures.
25:  // Applies a mutable lambda with borrowed captures.

Tests/ManglingTests/ManglingTests.swift
51:        let lambda = fun[let capture = local](x: Int) { local + x }
53:          let z = lambda(x: y)

Tests/LibraryTests/TestCases/PointerToMutableTests.hylo
27:fun test_initialize_pointee_lambda() {
46:  test_initialize_pointee_lambda()
```
</details>

### CODE REVIEW

File: Sources/FrontEnd/TypeChecking/TypeChecker.swift [hunk 1]
```diff
@@ -2543,16 +2543,16 @@ struct TypeChecker {
   }
 
   /// Computes and returns the types of the inputs of `e`'s underlying declaration, using `hint`
-  /// to guess the passing conventions of unnaotated parameters.
+  /// to guess the passing conventions of unanotated parameters.
   ///
-  /// `hint` is used as contextual information to refine guesses iff it is a lambda type with the
+  /// `hint` is used as contextual information to refine guesses if it is a lambda type with the
```
Comment (kyouko-taiga) [inline Sources/FrontEnd/TypeChecking/TypeChecker.swift:2548]: `iff` is not a typo. We use it to mean "if and only if", which is a double implication.

File: Sources/FrontEnd/TypeChecking/TypeChecker.swift [hunk 2]
```diff
@@ -2887,8 +2887,8 @@ struct TypeChecker {
 
     if program[e].isAutoclosure {
       let s = program[program[e].bareType].site
-      guard let u = LambdaType(t), u.inputs.isEmpty else {
-        report(.error(autoclosureExpectsEmptyLambdaAt: s, given: t))
+      guard let u = ArrowType(t), u.inputs.isEmpty else {
+        report(.error(autoclosureExpectsEmptyArrowAt: s, given: t))
```
Comment (kyouko-taiga) [inline Sources/FrontEnd/TypeChecking/TypeChecker.swift:2891]: I think the deserve some work but FWIW it does refer to an actual lambda. Maybe `autoclosureExpectsEmptyEnvironment`.

File: Sources/IR/Callee.swift [hunk 3]
```diff
@@ -9,8 +9,8 @@ enum Callee {
   /// A reference to a variant in a method bundle.
   case bundle(BundleReference<MethodDecl>)
 
-  /// A lambda.
-  case lambda(Operand)
+  /// An arrow.
+  case arrow(Operand)
```
Comment (kyouko-taiga) [inline Sources/IR/Callee.swift:13]: That would be an actual lambda.

**Review by kyouko-taiga — COMMENTED** [review]
Thanks for this PR! I'm already riding the "arrow" train but seeing the change in context, I really think that is a big improvement.

I left comments for a couple of changes that should not be made.

File: Sources/FrontEnd/TypeChecking/TypeChecker.swift (hunk 2 above)
Comment (peter-evans) [reply Sources/FrontEnd/TypeChecking/TypeChecker.swift]: https://github.com/hylo-lang/hylo/pull/1293/commits/93713afe5b6b00a276241b45f18971114c94ea1f

File: Sources/IR/Callee.swift (hunk 3 above)
Comment (peter-evans) [reply Sources/IR/Callee.swift]: https://github.com/hylo-lang/hylo/pull/1293/commits/92ac1dee8711669945decffa9a71d4016a8130b1

File: Sources/Core/Types/ArrowType.swift [hunk 4]
```diff
@@ -1,18 +1,18 @@
 import Utils
 
-/// The type of a lambda.
-public struct LambdaType: TypeProtocol {
+/// A type denoting arrow-like objects.
```
Comment (kyouko-taiga) [inline Sources/Core/Types/ArrowType.swift:3]:
```suggestion
/// A type expressed with an arrow.
```

**Review by kyouko-taiga — APPROVED** [review]
OK for me. "arrow" may not have convinced every one but regardless I think it is fair to say this PR is an improvement over the status quo. We can always rename instances of "ArrowType" by "WhateverType" if we get more inspiration.

**Pull request merged**